### PR TITLE
[aotinductor] Pass TorchIR to AOTInductor

### PR DESCRIPTION
**Pull request by angelayi**
Updates `_export.aot_compile` to pass a torch IR graph to inductor, allowing inductor to now run the pre_grad_passes, and reuse more of inductor's code. 
Also updates the API to only return the `so_path`, and not returning the exported program. The pytree call spec is now serialized and placed inside of the generated model code. When calling the model, because there is no c++ pytree implementation linked yet, we can access the call specs through `get_call_spec()`, and call pytree flatten/unflattenin python.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan

### CODE REVIEW

File: benchmarks/dynamo/common.py [hunk 1]
```diff
@@ -1200,15 +1192,21 @@ def opt_export(_, example_inputs):
 
 
 def export_aot_inductor(model, example_inputs):
-    module, exported = AOTInductorModelCache.load(model, example_inputs)
+    module = AOTInductorModelCache.load(model, example_inputs)
+    call_spec = module.get_call_spec()
+    call_spec = (
```
Comment (desertfire) [inline benchmarks/dynamo/common.py:1197]: Better to assign to two variables with explicit naming

File: test/inductor/test_aot_inductor.py [hunk 2]
```diff
@@ -58,18 +59,20 @@ class AOTInductorModelRunner:
...
             remove_runtime_assertions=True,
         )
-        return so_path, exported
+        return so_path
 
     @classmethod
     def load(cls, so_path, example_inputs):
-        is_cpu = all(x.device.type == "cpu" for x in example_inputs)
+        is_cpu = all(
+            x.device.type == "cpu" for x in pytree.tree_flatten(example_inputs)[0]
```
Comment (desertfire) [inline test/inductor/test_aot_inductor.py:74]: Are you only checking the first input now?

File: test/inductor/test_aot_inductor.py [hunk 3]
```diff
@@ -81,33 +84,36 @@ def load(cls, so_path, example_inputs):
...
-                functions=["run"],
+                functions=["run", "get_call_spec"],
                 extra_ldflags=[so_path],
                 with_cuda=not is_cpu,
-            ).run
+            )
 
-        return optimized
+            call_spec = module.get_call_spec()
+            call_spec = (
```
Comment (desertfire) [inline test/inductor/test_aot_inductor.py:98]: Similar suggestion for the naming here

File: torch/_inductor/codegen/aoti_runtime/interface.cpp [hunk 4]
```diff
@@ -129,6 +129,19 @@ AOTIRuntimeError AOTInductorModelContainerGetOutputName(
       { *ret_output_names = container->output_name(output_idx); })
 }
 
+AOTIRuntimeError AOTInductorModelContainerGetCallSpec(
+    AOTInductorModelContainerHandle container_handle,
+    std::string& in_spec,
```
Comment (desertfire) [inline torch/_inductor/codegen/aoti_runtime/interface.cpp:134]: These interface calls should only use C. Let's change it to `char *`.

File: test/inductor/test_aot_inductor.py (hunk 2 above)
Comment (angelayi) [reply test/inductor/test_aot_inductor.py:74]: pytree.tree_flatten returns a flattened list of example_inputs and the tree_spec for how it flattened the original list. So indexing [0] just gets the entire flattened list.

**Review by desertfire — APPROVED** [review]

**Pull request closed without merging**